Fixes #31992 - Registering hosts with enabled FIPS

stejskalleos opened the pull request:
<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

stejskalleos committed:
Fixes #31992 - Registration of host with FIPS

Workaround for systems with enabled FIPS,
where installation of RPM generated on RHEL7 cause 'no digest' error.
For more details see:
- https://projects.theforeman.org/issues/32068
- https://access.redhat.com/solutions/4460971

ekohl commented on app/views/unattended/provisioning_templates/registration/global_registration.erb:
```diff
@@ -72,7 +72,13 @@ if [ x$ID = xrhel ] || [ x$ID = xcentos ]; then
     curl --silent --output $CONSUMER_RPM <%= subscription_manager_configuration_url(hostname: @url_host) %>
-    yum localinstall $CONSUMER_RPM -y
+
+    if [ "$(cat /proc/sys/crypto/fips_enabled)" = "1" ]; then
+        rpm -ivh --nodigest --nofiledigest $CONSUMER_RPM
```
This deserves a comment because it's a workaround. Normally we open an issue for that and link it in the code. It probably means we should start generating a consumer RPM with a secure digest instead of simply not checking it.

That also makes me doubt if we even need this in the first place or rather we should fix the source.

ekohl commented:
@ehelms I'm leaning to closing this and fixing our consumer RPM to be signed with SHA256. Do you see any issues with that?

ehelms commented:
We do not currently sign the consumer RPM. What aspect of the RPM generation are we meaning here with the digest? 

I see, from 2015, a change to sha256: https://github.com/Katello/katello-certs-tools/commit/b68836ab1b70d085691168dbc3748769c405e522

ekohl commented:
No, it's the digest in the RPM. https://www.starlab.io/blog/adding-sha256-digests-to-rpms suggests that to me at least though I didn't check it myself.

stejskalleos commented:
Rebased & updated
I added comment explaining the workaround with the link to the issue with rpm digest (https://projects.theforeman.org/issues/32068)

ekohl reviewed (changes requested):
On Katello 3.18 EL7:
```console
$ rpm --verbose --checksig /var/www/html/pub/katello-ca-consumer-latest.noarch.rpm
/var/www/html/pub/katello-ca-consumer-latest.noarch.rpm:
    Header SHA1 digest: OK (3afd39d9dac6066f5d832972733f8c8bae2b6819)
    MD5 digest: OK (b170d40dfe4fe50500253d2f0f848a55)
```

This is NOT OK for EL8 agents with FIPS enabled.

On Katello nightly EL8 from when it was 4.0:
```console
$ rpm --verbose --checksig /var/www/html/pub/katello-ca-consumer-latest.noarch.rpm
/var/www/html/pub/katello-ca-consumer-latest.noarch.rpm:
    Header SHA256 digest: OK
    Header SHA1 digest: OK
    Payload SHA256 digest: OK
```

This is OK for FIPS. The critical part here is that EL8 signs with SHA256 while EL7 doesn't. IMHO that should be fixed and this PR closed. Disabling digests on a more secure platform is a bad idea and actually reduces integrity.

I'm now working on a fix for the installer to see if we can teach it to generate SHA256 digests.

ekohl commented on app/views/unattended/provisioning_templates/registration/global_registration.erb:
```diff
@@ -72,7 +72,15 @@ if [ x$ID = xrhel ] || [ x$ID = xcentos ]; then
     curl --silent --output $CONSUMER_RPM <%= subscription_manager_configuration_url(hostname: @url_host) %>
-    yum localinstall $CONSUMER_RPM -y
+
+    # Workaround for 'no digest' error on systems with enabled FIPS
```
This should also note that it only affects consumer RPMs generated on EL < 8.

ekohl reviewed (commented):
It took me longer to verify this because I didn't read the Redmine issue. https://access.redhat.com/solutions/4460971. I once more would like to emphasize that if you want a smooth review, it really helps to provide all of this information in the commit itself. That can be in the commit message, in the code or both. That is also what will mostly be read in the future (because git blame makes that very easy).

It looks like we can't generate a correct RPM on EL7 so if you add the note that it only affects RPMs generated on EL < 8 I'm ok with merging this.

Bonus points if you provide a short summary in the commit message. https://chris.beams.io/posts/git-commit/ is the resource I recommend to everyone because it's awesome.

stejskalleos commented:
@ekohl Rebased & updated

ekohl reviewed (approved):
Thanks!
